Fix: Data Race Bug detected via TestConcurrenyDownloadIndex

haojingcn opened the pull request:
**What this PR does / why we need it**:
close #31536 

**Special notes for your reviewer**:
Add a RWMutex in struct ChartRepository, for avoid client get data race within concurrency.

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

haojingcn committed:
Fix: Data Race Bug detected via TestConcurrenyDownloadIndex

Signed-off-by: haojingcn <mhjlq1989@gmail.com>

TerryHowe reviewed (commented):
/lgtm

gjenkins8 reviewed (changes requested):
The point of `TestConcurrenyDownloadIndex` (sic) as far as I can tell to show that concurrent actions on index files are correct.

>	// Simultaneously start multiple goroutines that:
	// 1) download index.yaml via DownloadIndexFile (write operation),
	// 2) read index.yaml via LoadIndexFile (read operation).
	// This checks for race conditions and ensures correct behavior under concurrent read/write access.

The change here serializes all calls `DownloadIndexFile` (for the given `ChartRepository` instance). Which is simply serializing the parallel calls to `DownloadIndexFile` from the test (may as well delete the test at this point)

Instead, IMHO the test should be amended to use a `ChartRepository` per go-routine, to avoid concurrent operation on the same `ChartRepository` instance. Which retains validating the concurrent index access, without concurrent `ChartRepository` access.

haojingcn committed:
Fix: Data Race Bug detected via TestConcurrenyDownloadIndex.

use a ChartRepository per go-routine, to avoid concurrent operation on the same ChartRepository instance.

Signed-off-by: haojingcn <mhjlq1989@gmail.com>

mrlunchbox777 reviewed (approved):
validated that these tests changes work and it keep concurrency by creating a `NewChartRepository` in each goroutine.
